pkg/agent/openflow/feature.go

```diff
@@ -0,0 +1,280 @@
+//
+// - featurePodConnectivity, implementation of connectivity for Pods, activated by default.
+// - featureNetworkPolicy, implementation of K8s NetworkPolicy and Antrea NetworkPolicy, activated by default.
+// - featureService, implementation of K8s Service, activated by default.
+// - featureEgress, implementation of Egress, activation is determined by feature gate Egress.
+// - featureMulticast, implementation of multicast, activation is determined by feature gate Multicast,
+// - featureTraceflow, implementation of Traceflow.
+type feature interface {
+	// getFeatureName returns the name of the feature.
+	getFeatureName() string
+	// getRequiredTables returns a slice of required tables of the feature.
+	getRequiredTables() []*Table
+	// initFlows returns the initial flows of the feature.
+	initFlows() []binding.Flow
+	// replayFlows returns the fixed and cached flows that need to be replayed after OVS is reconnected.
+	replayFlows() []binding.Flow
+}
+
+const (
+	// Pipeline is used to implement a major function in Antrea data path. At this moment, we have the following pipelines:
+
+	// pipelineRoot is only used to initialize PipelineClassifierTable.
+	pipelineRoot binding.PipelineID = iota
+	// pipelineARP is used to process ARP packets.
+	pipelineARP
+	// pipelineIP is used to process IPv4/IPv6 packets
+	pipelineIP
+	// pipelineMulticast is used to process multicast packets.
+	pipelineMulticast
+)
+
+const (
+	// Stage is used to group tables which implement similar functions in a pipeline. At this moment, we have the following
+	// stages:
+
+	// stageStart is only used to initialize PipelineClassifierTable.
+	stageStart binding.StageID = iota
+	// Classify packets "category" (tunnel, local gateway or local Pod, etc).
+	stageClassifier
+	// Validate packets.
+	stageValidation
+	// Transform committed packets in CT zones.
+	stageConntrackState
+	// Similar to PREROUTING chain of nat table in iptables. DNAT for Service connections is performed in this stage.
+	stagePreRouting
+	// Install egress rules for K8s NetworkPolicy and Antrea NetworkPolicy.
+	stageEgressSecurity
+	// L3 Forwarding of packets.
+	stageRouting
+	// Similar to POSTROUTING chain of nat table in iptables. SNAT for Service connections is performed in this stage.
+	stagePostRouting
+	// L2 Forwarding of packets.
+	stageSwitching
+	// Install ingress rules for K8s NetworkPolicy and Antrea NetworkPolicy.
+	stageIngressSecurity
+	// Commit non-Service connections.
+	stageConntrack
+	// Output packets to target port.
+	stageOutput
+)
+
+// Table in FlexiblePipeline is the basic unit to build OVS pipelines. A Table can be referenced by one or more features,
+// but its member struct ofTable will be initialized and realized on OVS only when it is referenced by any activated features.
+type Table struct {
+	name     string
+	stage    binding.StageID // All stages are defined in pkg/ovs/openflow/interfaces.go
```

Comment:
The comment is out-of-date? or you should move the pipelines and stages there to interfaces.go? Do we plan to share pipelines and stages for different use cases or not?

Reply:
Sorry, forget removing this.